src/format/mod.rs

```diff
@@ -0,0 +1,888 @@
+    WeekDayFrom0,
+    /// `"%G"`: Same as `%Y`, but using the ISO 8601 week-based year.
+    YearIso8601,
+    /// `"%g"`: Same as `%y`, but using the ISO 8601 week-based year.
+    YearIso8601Rem100,
+    /// `"%V"`: ISO 8601 week number in `01..=53`, zero-padded to 2 digits.
+    WeekNumberIso8601,
+    /// `"%U"`: Week number from Sunday in `00..=53`, zero-padded to 2 digits. The week `1` starts with the first Sunday of the year.
+    WeekNumberFromSunday,
+    /// `"%W"`: Week number from Monday in `00..=53`, zero-padded to 2 digits. The week `1` starts with the first Monday of the year.
+    WeekNumberFromMonday,
+    /// `"%s"`: Number of seconds since `1970-01-01 00:00:00 UTC`, zero-padded to at least 1 digit.
+    SecondsSinceEpoch,
+    /// `"%n"`: Newline character `'\n'`.
+    Newline,
+    /// `"%t"`: Tab character `'\t'`.
+    Tabulation,
+    /// `"%%"`: Literal `'%'` character.
+    Percent,
+    /// `"%c"`: Date and time, equivalent to `"%a %b %e %H:%M:%S %Y"`.
+    CombinationDateTime,
+    /// `"%D"` and `"%x"`: Date, equivalent to `"%m/%d/%y"`.
+    CombinationDate,
+    /// `"%F"`: ISO 8601 date, equivalent to `"%Y-%m-%d"`.
+    CombinationIso8601,
+    /// `"%v"`: VMS date, equivalent to `"%e-%^b-%4Y"`.
+    CombinationVmsDate,
+    /// `"%r"`: 12-hour time, equivalent to `"%I:%M:%S %p"`.
+    CombinationTime12h,
+    /// `"%R"`: 24-hour time without seconds, equivalent to `"%H:%M"`.
+    CombinationHourMinute24h,
+    /// `"%T"` and `"%X"`: 24-hour time, equivalent to `"%H:%M:%S"`.
+    CombinationTime24h,
+}
+
+/// UTC offset parts.
+#[derive(Debug)]
+struct UtcOffset {
+    /// Signed hour.
+    hour: f64,
+    /// Minute.
+    minute: u32,
+    /// Second.
+    second: u32,
+}
+
+impl UtcOffset {
+    /// Construct a new `UtcOffset`.
+    fn new(hour: f64, minute: u32, second: u32) -> Self {
+        Self {
+            hour,
+            minute,
+            second,
+        }
+    }
+}
+
+/// Formatting directive.
+#[derive(Debug)]
+struct Piece {
+    /// Optional width.
+    width: Option<usize>,
+    /// Padding method.
+    padding: Padding,
+    /// Formatting flags.
+    flags: Flags,
+    /// Formatting specifier.
+    spec: Spec,
+}
+
+impl Piece {
+    /// Construct a new `Piece`.
+    fn new(width: Option<usize>, padding: Padding, flags: Flags, spec: Spec) -> Self {
+        Self {
+            width,
+            padding,
+            flags,
+            spec,
+        }
+    }
+
+    /// Format a numerical value, padding with zeros by default.
+    fn format_num_zeros(
+        &self,
+        f: &mut SizeLimiter<'_>,
+        value: impl fmt::Display,
+        default_width: usize,
+    ) -> Result<(), Error> {
+        if self.flags.contains(Flags::LEFT_PADDING) {
+            write!(f, "{}", value)
+        } else if self.padding == Padding::Spaces {
+            let width = self.width.unwrap_or(default_width);
+            write!(f, "{: >width$}", value)
```

Comment:
whoa the `write!` macro works with the custom `SizeLimiter` type based on `crate::write::Write`? I did not realize it was that polymorphic.

Reply:
Yes it works as long as you have a `write_fmt` method taking a `core::fmt::Arguments` as input.